TRE core should use private endpoint to access TRE management storage account #4353

ashis-kar91 opened the pull request:
# Resolves #4353

## What is being addressed

- Access to TRE _management_ storage account was allowed from all networks, which is a security risk. TRE _core_ used to access the _management_ storage account via public network.
- TRE _management_ storage account should use a _private endpoint_ to restrict public network access and ensure data remains within a private network.
- Local TRE deployments will fail by making the storage account completely private. Hence _Allow public access from specific IP addresses_ during deployment.

## How is this addressed

The changes include the addition of a private endpoint for the TRE management storage account from the resource processor subnet, updates to network rules for the storage account to deny public access, and the introduction of a new script to handle adding and removing network exceptions for the management storage account during deployments.

### Security Enhancements:
* `core/terraform/resource_processor/vmss_porter/main.tf`: Add a new resource `azurerm_private_endpoint` for the TRE management storage account to use a private endpoint. This ensures secure access to the management storage account from TRE core resource processor subnet.
* Set default action to "Deny" and bypass to "AzureServices" for the TRE management storage account (`devops/terraform/bootstrap.sh`) [[1]](diffhunk://#diff-b8141bce138b6f8ec9bbe6df18ed9c586e940032a8ca17f6151dc943d4a9b0acR23-R24) [[2]](diffhunk://#diff-b8141bce138b6f8ec9bbe6df18ed9c586e940032a8ca17f6151dc943d4a9b0acR33-R35).
* Add network rules to deny public access and allow only specific IPs for the TRE management storage account (`devops/terraform/main.tf`).

### Deployment Script Improvements:
* Add script `mgmtstorage_add_network_exception.sh` to handle adding and removing network exceptions for the management storage account (`devops/scripts/mgmtstorage_add_network_exception.sh`).
* Update deployment scripts to source the new `mgmtstorage_add_network_exception.sh` script (`core/terraform/deploy.sh`, `core/terraform/destroy.sh`, `devops/terraform/deploy.sh`, `devops/terraform/destroy.sh`) [[1]](diffhunk://#diff-5594d7f9af1dd1963cb6c49f03744f3660683cd123abb715eb9bc725e432b9a6R11-R13) [[2]](diffhunk://#diff-68fdad88c917b9d8fee4eeae617a38e731347dd739a87490d2d18aba12a0afb8R11-R13) [[3]](diffhunk://#diff-16780962bc2fc3ddf7932225062b4cfbd35613f2e1c25f180f8a5c7fb15c7413R8-R10) [[4]](diffhunk://#diff-e1497f01c0603771a756bdbb02f9da64fafc182c7cf777d9fd787d494dc0c614R1-R10).

### Version Updates:
* Update version information in `core/version.txt` and `devops/version.txt` [[1]](diffhunk://#diff-833c80f270ad0fc1102a2689ded40bd99f5fa1abdcb882420ebf09a326d84f27L1-R1) [[2]](diffhunk://#diff-fbaa85954bf392d1ceeb70ec7b34398613dd1a18ef328ef97b2bbf172e3ebd22L1-R1).

### Changelog update:
* Deny public access to the TRE management storage account and add private endpoint for TRE core (`CHANGELOG.md`).

marrobi commented:
@ashis-kar91 thanks, I might be confused, but when deploy say, using IP 1.2.3.4, first time, that IP gets added to the firewall for the state storage account. Next time, the agent has 6.7.8.9, how does the agent access the terraform state store?

I think the storage firewall needs to be amended using the CLI prior to the terraform commands running, similar to #4260 with the KeyVault.

@jonnyry do you concur?

ashis-kar91 commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 **BREAKING CHANGES & MIGRATIONS**:
 
 ENHANCEMENTS:
+* TRE core should use private endpoint to access TRE management storage account ([#4353](https://github.com/microsoft/AzureTRE/issues/4353))
```
I don't think we should make it `Change TRE management storage account network access from public to disabled` as we are not disabling it completely, instead we are allowing it from specific local IPs.
`Deny public access to TRE management storage account and add private endpoint for TRE core` seems more appropriate. What do you think?

yuvalyaron replied:
Sounds good

ashis-kar91 commented on core/terraform/locals.tf:
```diff
@@ -13,6 +13,8 @@ locals {
 
   docker_registry_server = data.azurerm_container_registry.mgmt_acr.login_server
 
+  mgmt_storage_account_id = data.azurerm_storage_account.mgmt_storage.id
```
I am seeing 2 different conventions here in the code base.
At some places `data` is directly used in the templates and at other places it's referred via `local`. So, followed the one which was more modular.

yuvalyaron replied:
IMO there is not added value for storing this in the locals in this case

marrobi reviewed (changes requested):
Putting request changes on, until are clear on https://github.com/microsoft/AzureTRE/pull/4360#issuecomment-2652336263

ashis-kar91 commented:
> @ashis-kar91 thanks, I might be confused, but when deploy say, using IP 1.2.3.4, first time, that IP gets added to the firewall for the state storage account. Next time, the agent has 6.7.8.9, how does the agent access the terraform state store?
> 
> I think the storage firewall needs to be amended using the CLI prior to the terraform commands running, similar to #4260 with the KeyVault.
> 
> @jonnyry do you concur?

Sounds good to me. Working on it.

jonnyry commented:
> @ashis-kar91 thanks, I might be confused, but when deploy say, using IP 1.2.3.4, first time, that IP gets added to the firewall for the state storage account. Next time, the agent has 6.7.8.9, how does the agent access the terraform state store?
> 
> I think the storage firewall needs to be amended using the CLI prior to the terraform commands running, similar to #4260 with the KeyVault.
> 
> @jonnyry do you concur?

Yes - since the storage account is holding tfstate terraform will need data plane access during the plan phase.  The first run should succeed, second and subsequent runs will fail - since the IP added on the previous run will not match the current GitHub runner's IP.

marrobi commented on core/terraform/storage.tf:
```diff
@@ -35,6 +35,32 @@ resource "azurerm_storage_account" "stg" {
+  name                = "pe-mgmt-blob-${var.tre_id}"
+  location            = azurerm_resource_group.core.location
+  resource_group_name = azurerm_resource_group.core.name
+  subnet_id           = module.network.shared_subnet_id
```
Agree. Makes sense.

ashis-kar91 commented:
@marrobi, @jonnyry,
Made the changes to dynamically add and remove network exceptions. Also resolved all the pending comments.
Please review again.

marrobi reviewed (approved):
LGTM, will let the tests check the operation.

marrobi commented:
/test

marrobi commented:
/test 1f41328

marrobi commented:
@ashis-kar91 tests have failed on the new script - https://github.com/microsoft/AzureTRE/actions/runs/13367453953

jonnyry commented on devops/scripts/mgmtstorage_add_network_exception.sh:
```diff
@@ -0,0 +1,139 @@
+
+  if [[ "$COUNT" -gt 0 ]]; then
+    # Step 2: Try accessing storage to confirm access
+    if az storage container list --account-name "$SA_NAME" --auth-mode login --output none 2>/dev/null; then
```
Should error output be hidden with 2>/dev/null?  Useful to log what has caused the failure.

ashis-kar91 replied:
Done

jonnyry commented on devops/scripts/mgmtstorage_add_network_exception.sh:
```diff
@@ -0,0 +1,139 @@
+  COUNT=$(az storage account network-rule list --resource-group "$RESOURCE_GROUP" --account-name "$SA_NAME" --query "length(ipRules[?ipAddressOrRange=='$MY_IP'])" --output tsv)
+
+  if [[ "$COUNT" -gt 0 ]]; then
+    # Step 2: Try accessing storage to confirm access
```
Due to the distributed nature of storage accounts, you might find that just listing containers is not enough to guarantee they are ready to use.

I received occasional intermittent failures on similar code, until I added an extra check (in a loop) to list blobs on each container:

`az storage blob list --container-name "$CONTAINER_NAME" --account-name "$STORAGE_ACCOUNT_NAME" --auth-mode login`

ashis-kar91 replied:
Done

jonnyry commented on devops/terraform/main.tf:
```diff
@@ -34,6 +34,12 @@ resource "azurerm_storage_account" "state_storage" {
   shared_access_key_enabled        = false
   local_user_enabled               = false
 
+  network_rules {
```
Suggest adding `public_network_access` to be explicit about the full network configuration

ashis-kar91 replied:
Done

jonnyry reviewed (commented):
See inline comments.

Also, do any of the other `core/terraform` shell scripts that access the backend tfstate container, require your add network exception script adding? E.g. `migrate.sh` or `outputs.sh`.  I haven't had chance to look at the flows, but I'm not sure all of them are called within the standard deployment flow.

ashis-kar91 commented:
/test 1f41328

marrobi commented:
> /test [1f41328](https://github.com/microsoft/AzureTRE/commit/1f413280acc57c7d6314a454c281c6109ceba3dc)

If you let us know when the comments have ben resolved, can run another test. We have a fixed budget on the CI subscription, hence the automated test should be used as a final gate. Thanks.

ashis-kar91 commented on devops/terraform/bootstrap.sh:
```diff
@@ -20,6 +20,8 @@ if ! az storage account show --resource-group "$TF_VAR_mgmt_resource_group_name"
     --name "$TF_VAR_mgmt_storage_account_name" --location "$LOCATION" \
     --allow-blob-public-access false --min-tls-version TLS1_2 \
     --kind StorageV2 --sku Standard_LRS -o table \
+    --default-action Deny \
```
Done

ashis-kar91 commented:
> See inline comments.
> 
> Also, do any of the other `core/terraform` shell scripts that access the backend tfstate container, require your add network exception script adding? E.g. `migrate.sh` or `outputs.sh`. I haven't had chance to look at the flows, but I'm not sure all of them are called within the standard deployment flow.

The `migrate.sh` was called withing the standard deploy, but `outputs.sh` wasn't. So, I moved the exception script inside of those and removed it from `deploy.sh` and `destroy.sh`. Also added the exception script within the `terraform_wrapper.sh` as it was also using the container.

ashis-kar91 commented:
@marrobi, @jonnyry 
I have resolved all the review comments. Please review again.
@marrobi, I have tested all the scripts where I have added the exception. So shall we run E2E test now.

marrobi commented:
/test

marrobi commented:
/test a905066

jonnyry reviewed (approved):
@ashis-kar91 LGTM, thanks for the changes :-)

marrobi commented:
/test

marrobi commented:
/test 25e661a

marrobi commented:
/test-destroy-env

marrobi commented:
/test 4abe2c9

ashis-kar91 commented:
/test 5dc89a4

ashis-kar91 commented:
Are we good to merge @marrobi, @jonnyry?

marrobi commented on CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@ BUG FIXES:
 * Workspace bundle uses infrastructure encryption on shared storage which will recreate storage share. Major verison increase will prevent upgrade, do not force the upgrade unless you are fully aware of the consequences.
 
 ENHANCEMENTS:
+* Deny public access to TRE management storage account, and add private endpoint for TRE core [#4353](https://github.com/microsoft/AzureTRE/issues/4353)
```
@ashis-kar91 can you move this to the section above?

ashis-kar91 replied:
Done

marrobi commented:
/test-force-approve 

Passed: https://github.com/microsoft/AzureTRE/actions/runs/13460780708
